gtsam/sfm/PowerMethod.h

```diff
@@ -0,0 +1,367 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010-2019, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file   PowerMethod.h
+ * @date   Sept 2020
+ * @author Jing Wu
+ * @brief  accelerated power method for fast eigenvalue and eigenvector computation
+ */
+
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Vector.h>
+#include <gtsam/dllexport.h>
+
+#include <Eigen/Core>
+#include <Eigen/Sparse>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <algorithm>
+#include <cmath>
+#include <chrono>
+#include <random>
+
+namespace gtsam {
+
+using Sparse = Eigen::SparseMatrix<double>;
+
+/* ************************************************************************* */
+/// MINIMUM EIGENVALUE COMPUTATIONS
+
+struct PowerFunctor {
+  /**
+   * \brief Computer i-th Eigenpair with power method
+   *
+   * References :
+   * 1) Rosen, D. and Carlone, L., 2017, September. Computational
+   * enhancements for certifiably correct SLAM. In Proceedings of the
+   * International Conference on Intelligent Robots and Systems.
+   * 2) Yulun Tian and Kasra Khosoussi and David M. Rosen and Jonathan P. How,
+   * 2020, Aug, Distributed Certifiably Correct Pose-Graph Optimization, Arxiv
+   * 3) C. de Sa, B. He, I. Mitliagkas, C. Ré, and P. Xu, “Accelerated
+   * stochastic power iteration,” in Proc. Mach. Learn. Res., no. 84, 2018, pp. 58–67
+   *
+   * It performs the following iteration: \f$ x_{k+1} = A * x_k + \beta *
+   * x_{k-1} \f$ where A is the certificate matrix, x is the ritz vector
+   *
+   */
+
+  // Const reference to an externally-held matrix whose minimum-eigenvalue we
+  // want to compute
+  const Sparse &A_;
+
+  const Matrix &S_;
+
+  const int m_n_; // dimension of Matrix A
+  const int m_nev_; // number of eigenvalues required
+
+  // flag for running power method or accelerated power method. If false, the former, vice versa.
+  bool accelerated_;
+
+  // a Polyak momentum term
+  double beta_;
+
+  // const int m_ncv_; // dimention of orthonormal basis subspace
+  size_t m_niter_; // number of iterations
+
+private:
+  Vector ritz_val_; // all ritz eigenvalues
+  Matrix ritz_vectors_; // all ritz eigenvectors
+  Vector ritz_conv_; // store whether the ritz eigenpair converged
+  Vector sorted_ritz_val_; // sorted converged eigenvalue
+  Matrix sorted_ritz_vectors_; // sorted converged eigenvectors
+
+public:
+ // Constructor
+ explicit PowerFunctor(const Sparse& A, const Matrix& S, int nev, int ncv,
```

Comment:
The fact that there's an A AND an S here tells me that this is NOT just doing a basic power method computation -- it looks like this is doing the full minimum eigenpair computation.

You should separate the (optionally accelerated) power method from the full minimum-eigenvalue computation by writing two separate functions.  This is logically cleaner/clearer, and it's better for code re-use and unit testing.

Reply:
The S is passed here because I want to use the first row of S as the initial x0 as mention in the III.C part of Computational Enhancements for Certifiably Correct SLAM paper. You're right, it's less confusing if I can pass only the first row.

Now the PowerFunctor is only doing the power method.